packaging/tags.py

```diff
@@ -162,91 +183,156 @@ def _cpython_abis(py_version, warn=False):
+
+    if abis is None:
+        if len(python_version) > 1:
+            abis = _cpython_abis(python_version, warn)
+        else:
+            abis = []
+    abis = list(abis)
+    # 'abi3' and 'none' are explicitly handled later.
+    for explicit_abi in ("abi3", "none"):
+        try:
+            abis.remove(explicit_abi)
+        except ValueError:
+            pass
+
+    platforms = list(platforms or _platform_tags())
     for abi in abis:
         for platform_ in platforms:
             yield Tag(interpreter, abi, platform_)
-    for tag in (Tag(interpreter, "abi3", platform_) for platform_ in platforms):
-        yield tag
+    # Not worrying about the case of Python 3.2 or older being specified and
+    # thus having redundant tags thanks to the abi3 in-fill later on as
+    # 'packaging' doesn't directly support Python that far back.
+    if len(python_version) > 1:
+        for tag in (Tag(interpreter, "abi3", platform_) for platform_ in platforms):
+            yield tag
     for tag in (Tag(interpreter, "none", platform_) for platform_ in platforms):
         yield tag
     # PEP 384 was first implemented in Python 3.2.
-    for minor_version in range(py_version[1] - 1, 1, -1):
-        for platform_ in platforms:
-            interpreter = "cp{major}{minor}".format(
-                major=py_version[0], minor=minor_version
-            )
-            yield Tag(interpreter, "abi3", platform_)
-
-
-def _pypy_interpreter():
-    # type: () -> str
-    # Ignoring sys.pypy_version_info for type checking due to typeshed lacking
-    # the reference to the attribute.
-    return "pp{py_major}{pypy_major}{pypy_minor}".format(
-        py_major=sys.version_info[0],
-        pypy_major=sys.pypy_version_info.major,  # type: ignore
-        pypy_minor=sys.pypy_version_info.minor,  # type: ignore
-    )
+    if len(python_version) > 1:
+        for minor_version in range(python_version[1] - 1, 1, -1):
+            for platform_ in platforms:
+                interpreter = "cp{major}{minor}".format(
+                    major=python_version[0], minor=minor_version
+                )
+                yield Tag(interpreter, "abi3", platform_)
 
 
 def _generic_abi():
-    # type: () -> str
+    # type: () -> Iterator[str]
     abi = sysconfig.get_config_var("SOABI")
     if abi:
-        return _normalize_string(abi)
-    else:
-        return "none"
+        yield _normalize_string(abi)
 
 
-def _pypy_tags(py_version, interpreter, abi, platforms):
-    # type: (PythonVersion, str, str, Iterable[str]) -> Iterator[Tag]
-    for tag in (Tag(interpreter, abi, platform) for platform in platforms):
-        yield tag
-    for tag in (Tag(interpreter, "none", platform) for platform in platforms):
-        yield tag
+def _generic_interpreter(warn=False):
+    # type: (bool) -> str
+    version = _get_config_var("py_version_nodot", warn)
+    if not version:
+        version = "".join(map(str, sys.version_info[:2]))
```

Comment:
Can we expose this as `interpreter_version_nodot()`? It would ease the construction of `interpreter` for `generic_tags` and `compatible_tags`, since we take implementation and version separately and fill in one or the other if not provided.

Reply:
We can expose it as `interpreter_version()` (the `nodot` is an implementation detail).